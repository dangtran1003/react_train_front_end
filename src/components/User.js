import React, { Component} from 'react';
import PropTypes from 'prop-types'
import {user_del} from '../actions/user'
import { connect } from 'react-redux';

class User extends Component {
  static propTypes = {
    username : PropTypes.string,
    name : PropTypes.string,
    email : PropTypes.string,
    id: PropTypes.number
  }

  render() {
    const {username, name, email, id} = this.props
    return (
        <tr>
        <td>{username}</td>
        <td>{name}</td>
        <td>{email}</td>
        <td style={{width: 100}}>
          <a href={"/modify/"+id.toString()} className="btn btn-info" role="button" style={{margin:4}}><i className="icons font-l d-block mt-0 cui-wrench"></i></a>
          <button className="btn btn-danger" value={id} onClick={this.props.user_del}><i className="icons font-l d-block mt-0 cui-trash"></i></button>
        </td>
      </tr>
    )
  }
}

// export default User

const mapDispatchToProps = {
  user_del: user_del,
}

export default connect(null, mapDispatchToProps)(User)