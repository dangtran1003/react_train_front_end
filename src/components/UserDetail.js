import React, { Component} from 'react';
import PropTypes from 'prop-types'
import { connect } from 'react-redux';
import { user_detail, change_detail } from '../actions/user';

export class UserDetail extends Component {
  static propTypes = {
    email : PropTypes.string,
    name : PropTypes.string,
  }
  state = {
      email: '',
      name : '',
      error : ''
  }

  componentWillMount() {
     this.props.UserDetail(this.props.match.params.id)
  }
  handleFormChange = e =>{
       const name = e.target.name
       const value = e.target.value
       this.setState({
        [name] :value
      }) 
   }
   handleSubmit = e =>
        {
            this.props.ChangeDetail(this.props.match.params.id, this.state.email, this.state.name)
        }

  render() {
    return (
        <div className="card">
            <div className="card-header">
                <strong>Cập nhật thông tin nhân viên <i>{this.props.user.username}</i></strong>
            </div>
            <div className="card-body">
                <form className="form-horizontal">
                    <div className="form-group row">
                        <label className="col-md-3 col-form-label" htmlFor="hf-email">Email</label>
                        <div className="col-md-9">
                        <input className="form-control"  type="email" name="email" onChange={this.handleFormChange} autoComplete="email" placeholder={this.props.user.email}/>
                        <span className="help-block">Please enter your email</span>
                        </div>
                    </div>
                    <div className="form-group row">
                        <label className="col-md-3 col-form-label" htmlFor="hf-password">Name</label>
                        <div className="col-md-9">
                            <input className="form-control" type="text" name="name" onChange={this.handleFormChange} placeholder={this.props.user.name}/>
                            <span className="help-block">Please enter your name</span>
                        </div>
                    </div>
                </form>
            </div>
            <div className="card-footer">
                <button className="btn btn-sm btn-primary" onClick={this.handleSubmit}>
                    <i className="fa fa-dot-circle-o"></i> Submit</button>
            </div>
        </div>
    )
  }
}

const mapStateToProps = (state) => ({
    user: state.data.user
  })

function mapDispatchToProps(dispatch) {
    return {
        UserDetail : (id) => {
            dispatch(user_detail(id))
        },
        ChangeDetail : (id, email, name) => {
            dispatch(change_detail(id, email, name))
        }
    }
}
  
export default connect(mapStateToProps, mapDispatchToProps)(UserDetail)