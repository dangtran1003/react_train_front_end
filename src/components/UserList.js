import React, {Component} from 'react'
// import { connect } from "react-redux";
import PropTypes from 'prop-types'
import '../App.css'
import User from './User'
import {user_list, user_add} from '../actions/user'
import { connect } from 'react-redux';

export class UserList extends Component{
    state = {
      username : '',
      email : '',
      name : '',
      error: '', 
      key_search: ''
    }

    handleSubmit = e => {
      // e.preventDefault()
      if (this.state.username.length !== 0 && this.state.email.length !== 0)
        this.props.UserAdd(this.state.username, this.state.email, this.state.name)
      else 
        {
          this.setState({error:'Thiếu username hoặc password!'})
          alert('Thiếu username hoặc password!')
        }
    }

    handleFormChange = e =>{
      const name = e.target.name
      const value = e.target.value
      this.setState({
        [name] :value
      })
      if (name === 'key_search')
      {
        this.props.UserList(value,this.props.match.params.page-1,10)
      }
    }
    static propTypes = {
        users : PropTypes.array,
        total_users: PropTypes.array
    }

    static defaultProps = {
      users : [],
      total_users: [1]
    }

    componentDidMount() {
        this.props.UserList(null, this.props.match.params.page - 1, 10)
    }


    render(){
        return (
          <div className="card">
          <div className="card-header">
            <i> Danh sách người dùng</i>
            </div>
           
            {this.props.error && alert(this.props.error)}
          <div className="card-body">
          <input className="form-control" type="text" name="key_search" onChange={this.handleFormChange} placeholder="Key search"/>
          <br></br>
            <table className="table table-responsive-sm table-bordered table-striped table-sm">
              <thead>
                <tr>
                  <th name='username'>Username</th>
                  <th name='name'>Name</th>
                  <th name='email'>Email</th>
                  <th name='options'>Options</th>
                </tr>
              </thead>
              <tbody>
                {
                   this.props.users.map((user, key) => {
                    const {id, username, email, name} = user
                    return (
                      <User username={username} email={email} name = {name} key={key} id={id}></User>
                    )
                  })
                }
                <tr>
                  <td><input className="form-control" name="username" type="text" placeholder="Enter your username" onChange={this.handleFormChange}></input></td>
                  <td><input className="form-control" name="name" type="text" placeholder="Enter your name" onChange={this.handleFormChange}></input></td>
                  <td><input className="form-control" name="email" placeholder="Enter your email" onChange={this.handleFormChange} type="email"></input></td>
                  <td>
                      <button href="#" className="btn btn-success" style={{margin:4}} onClick={this.handleSubmit}><i className="icons font-l d-block mt-0 cui-check" ></i></button>
                      <button href="#" className="btn btn-danger"><i className="icons font-l d-block mt-0 cui-trash"></i></button>
                  </td>
                </tr>
              </tbody>
            </table>
            <nav>
              <ul className="pagination">
                <li className="page-item">
                  <a className="page-link" href={'list/'+(Number(this.props.match.params.page)>1 ? (Number(this.props.match.params.page)-1).toString() : "") }>Prev</a>
                </li>
                <li className={"page-time" +(this.props.match.params.page === 1 ? " active" : "") }>
                  <a className="page-link" href="list/1">1</a>
                </li>
                {this.props.pages.map((page,key)=>{
                  return (
                    <li className={"page-time" +(this.props.match.params.page === page ? " active" : "") } key = {key}>
                  <a className="page-link" href={'list/' + page.toString()}>{page}</a>
                </li>
                  )
                })}
                <li className="page-item">
                  <a className="page-link" href={'list/'+(Number(this.props.match.params.page) +1).toString()}>Next</a>
                </li>
              </ul>
            </nav>
          </div>
        </div>
        )
    }
}

const mapStateToProps = (state) => ({
  users: state.data.users,
  pages: state.data.pages,
  error: state.data.error
})

function mapDispatchToProps(dispatch) {
  return {
   UserList: (key_search,numPages,numUsers) => {
    dispatch(user_list(key_search, numPages,numUsers))
   },
   
   UserAdd : (username, email, name) => {
     dispatch(user_add(username, email, name))
   }
  }
 }


export default connect(mapStateToProps, mapDispatchToProps)(UserList)