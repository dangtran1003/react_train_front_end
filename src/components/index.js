import {connect} from 'react-redux'
import UserList from './UserList'

export default connect(({users}) => ({users}))(UserList)