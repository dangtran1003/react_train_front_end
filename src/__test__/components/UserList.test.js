// src/__test__/components/Link.test.js
import {shallow, mount} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import {UserList} from '../../components/UserList'
import User from '../../components/User'
import React from 'react'




const props = {
    users : [{username: "asasassssss", email: "sasassss", name: "ddang", id: 77},
    {username: "1111", email: "2222", name: "3333", id: 77}],
    pages : [1],
    error:'',
    match : { params : { page : 1}},
    UserList : jest.fn(),
    UserAdd : jest.fn(),
    user_del : jest.fn()
};

const state = {
    username : '',
    email : '',
    name : '',
    error: '', 
    key_search: ''
  }

describe('Components > UserList', () => {
    // Snapshot use renderer
   
    it('Render a snapshot for Link use enzyme', () => {
        window.alert = jest.fn()
        //fake event onChange input
        const event = {
          preventDefault() {},
          target: { value: 'the-value' }
        };
        const tree = shallow(<UserList {...props}/>);
        tree.instance().handleSubmit = jest.fn()
        tree.instance().handleFormChange = jest.fn()
        //kiem tra so User
        expect(tree.find('input')).toHaveLength(4)
        //test input form
        expect(tree.find('input').first().props().placeholder).toEqual('Key search')
        tree.find('input').first().simulate('change', event)
        expect(tree.instance().handleFormChange).toHaveBeenCalledTimes(0)
        expect(tree.find('input').at(1).props().placeholder).toEqual('Enter your username')
        tree.find('input').at(1).simulate('change', event)
        expect(tree.instance().handleFormChange).toHaveBeenCalledTimes(1)
        expect(tree.find('input').at(2).props().placeholder).toEqual('Enter your name')
        tree.find('input').at(2).simulate('change', event)
        expect(tree.instance().handleFormChange).toHaveBeenCalledTimes(2)
        expect(tree.find('input').at(3).props().placeholder).toEqual('Enter your email')
        tree.find('input').at(3).simulate('change', event)
        expect(tree.instance().handleFormChange).toHaveBeenCalledTimes(3)
        //test component User
        expect(tree.find(User)).toHaveLength(2)
        expect(tree.find(User).at(1).props()).toEqual(props.users[1])
        expect(tree.find(User).at(0).props()).toEqual(props.users[0])
        //test table header
        expect(tree.find('th')).toHaveLength(4)
        expect(tree.find('th').at(0).text()).toBe('Username')
        expect(tree.find('th').at(1).text()).toBe('Name')
        expect(tree.find('th').at(2).text()).toBe('Email')
        expect(tree.find('th').at(3).text()).toBe('Options')
        //test handle button
        //test button them user
        tree.find('button.btn-success').simulate('click')
        expect(tree.instance().handleSubmit).toHaveBeenCalledTimes(1);
        //match Snapshot
        expect(shallowToJson(tree)).toMatchSnapshot();
    })
})
