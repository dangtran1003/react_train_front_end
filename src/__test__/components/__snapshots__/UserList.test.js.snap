// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components > UserList Render a snapshot for Link use enzyme 1`] = `
<div
  className="card"
>
  <div
    className="card-header"
  >
    <i>
       Danh sách người dùng
    </i>
  </div>
  <div
    className="card-body"
  >
    <input
      className="form-control"
      name="key_search"
      onChange={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "preventDefault": [Function],
                "target": Object {
                  "value": "the-value",
                },
              },
            ],
            Array [
              Object {
                "preventDefault": [Function],
                "target": Object {
                  "value": "the-value",
                },
              },
            ],
            Array [
              Object {
                "preventDefault": [Function],
                "target": Object {
                  "value": "the-value",
                },
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        }
      }
      placeholder="Key search"
      type="text"
    />
    <br />
    <table
      className="table table-responsive-sm table-bordered table-striped table-sm"
    >
      <thead>
        <tr>
          <th
            name="username"
          >
            Username
          </th>
          <th
            name="name"
          >
            Name
          </th>
          <th
            name="email"
          >
            Email
          </th>
          <th
            name="options"
          >
            Options
          </th>
        </tr>
      </thead>
      <tbody>
        <Connect(User)
          email="sasassss"
          id={77}
          key="0"
          name="ddang"
          username="asasassssss"
        />
        <Connect(User)
          email="2222"
          id={77}
          key="1"
          name="3333"
          username="1111"
        />
        <tr>
          <td>
            <input
              className="form-control"
              name="username"
              onChange={
                [MockFunction] {
                  "calls": Array [
                    Array [
                      Object {
                        "preventDefault": [Function],
                        "target": Object {
                          "value": "the-value",
                        },
                      },
                    ],
                    Array [
                      Object {
                        "preventDefault": [Function],
                        "target": Object {
                          "value": "the-value",
                        },
                      },
                    ],
                    Array [
                      Object {
                        "preventDefault": [Function],
                        "target": Object {
                          "value": "the-value",
                        },
                      },
                    ],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                }
              }
              placeholder="Enter your username"
              type="text"
            />
          </td>
          <td>
            <input
              className="form-control"
              name="name"
              onChange={
                [MockFunction] {
                  "calls": Array [
                    Array [
                      Object {
                        "preventDefault": [Function],
                        "target": Object {
                          "value": "the-value",
                        },
                      },
                    ],
                    Array [
                      Object {
                        "preventDefault": [Function],
                        "target": Object {
                          "value": "the-value",
                        },
                      },
                    ],
                    Array [
                      Object {
                        "preventDefault": [Function],
                        "target": Object {
                          "value": "the-value",
                        },
                      },
                    ],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                }
              }
              placeholder="Enter your name"
              type="text"
            />
          </td>
          <td>
            <input
              className="form-control"
              name="email"
              onChange={
                [MockFunction] {
                  "calls": Array [
                    Array [
                      Object {
                        "preventDefault": [Function],
                        "target": Object {
                          "value": "the-value",
                        },
                      },
                    ],
                    Array [
                      Object {
                        "preventDefault": [Function],
                        "target": Object {
                          "value": "the-value",
                        },
                      },
                    ],
                    Array [
                      Object {
                        "preventDefault": [Function],
                        "target": Object {
                          "value": "the-value",
                        },
                      },
                    ],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                }
              }
              placeholder="Enter your email"
              type="email"
            />
          </td>
          <td>
            <button
              className="btn btn-success"
              href="#"
              onClick={
                [MockFunction] {
                  "calls": Array [
                    Array [],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                }
              }
              style={
                Object {
                  "margin": 4,
                }
              }
            >
              <i
                className="icons font-l d-block mt-0 cui-check"
              />
            </button>
            <button
              className="btn btn-danger"
              href="#"
            >
              <i
                className="icons font-l d-block mt-0 cui-trash"
              />
            </button>
          </td>
        </tr>
      </tbody>
    </table>
    <nav>
      <ul
        className="pagination"
      >
        <li
          className="page-item"
        >
          <a
            className="page-link"
            href="list/"
          >
            Prev
          </a>
        </li>
        <li
          className="page-time active"
        >
          <a
            className="page-link"
            href="list/1"
          >
            1
          </a>
        </li>
        <li
          className="page-time active"
          key="0"
        >
          <a
            className="page-link"
            href="list/1"
          >
            1
          </a>
        </li>
        <li
          className="page-item"
        >
          <a
            className="page-link"
            href="list/2"
          >
            Next
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
